packages/common/src/eos/ipfs-actions.ts

```diff
@@ -0,0 +1,103 @@
+import * as eosjsApi from "eosjs/dist/eosjs-api";
+import fs from "fs";
+import IpfsHash from "ipfs-only-hash";
+
+export interface ValidUploadActions {
+    [contract: string]: {
+        [action: string]: { maxSize: number; cidField: string };
+    };
+}
+
+export interface ActionIpfsData {
+    cid: string;
+    contract: string;
+    action: string;
+}
+
+export const parseActionIpfsCid = async (
+    serializedTransaction: Uint8Array,
+    validUploadActions: ValidUploadActions,
+    eosApi: eosjsApi.Api
+): Promise<ActionIpfsData> => {
+    const trx = eosApi.deserializeTransaction(serializedTransaction);
+
+    const serializedAction = trx.actions.find(
+        (action: any) =>
+            validUploadActions[action.account] &&
+            validUploadActions[action.account][action.name]
+    );
+
+    if (!serializedAction) {
+        throw new Error("contract action is not whitelisted for upload");
+    }
+
+    const actionData = (await eosApi.deserializeActions([serializedAction]))[0];
+
+    const contract = serializedAction.account;
+    const action = serializedAction.name;
+    const cid = extractCidFromActionData(
+        contract,
+        action,
+        actionData.data,
+        validUploadActions
+    );
+    return { cid, action, contract };
+};
+
+export const extractCidFromActionData = (
+    contract: string,
+    action: string,
+    data: any,
+    validUploadActions: ValidUploadActions
+) => {
+    const validContractActions = validUploadActions[contract];
+    if (!validContractActions) {
+        throw new Error(
+            `contract ${contract} has no valid actions for extracting cid`
+        );
+    }
+
+    const validAction = validContractActions[action];
+    if (!validAction) {
+        throw new Error(
+            `${contract}::${action} is not a valid action for extracting cid`
+        );
+    }
+
+    try {
+        // Extract the CID supporting dot notation fields, eg: new_member_profile.img
+        const cidFields = validAction.cidField.split(".");
+        let cid = data;
+        while (cidFields.length) {
+            const cidField = cidFields.shift()!;
+            cid = cid[cidField];
+        }
+        return cid;
```

Comment:
I find this loop to be very hard to understand. Likely just that I don't know what the data looks like, but maybe you add a comment that explains or shows a tiny sample of data? I'm guessing cidFields is a dot-separated list of properties of an object, and this function is using cidFields to drill into the object to get the cid, which is nested? Some way of making that easier to get than re-reading this code would be great.

Reply:
that's exactly it. just added the comment! done! 👍